Document style conventions for qualifying modules (#7378)
ocaml/dune: doc/hacking.rst

>   and are annoying to grep. Further more, they encourage callers not to think

Reviewer: ```suggestion
  and are annoying to grep. Furthermore, they encourage callers not to think
```

  at all about these optional arguments even if they often should.

- Avoid qualifying modules when accessing fields of records or constructors.

>   Prefer to avoid it altogether if possible or add a type annotation if

Reviewer: ```suggestion
  Avoid it altogether if possible, or add a type annotation if
```

  necessary.

Bad:

.. code:: ocaml

    let result = A.b () in
    match result.A.field with

>     | B.Constructor -> ...

Reviewer 2: There is an OCaml warning for this no? Maybe we should use it. Is it 40?

Author: It seems like the warning is enabled by default but dune disables it. Unfortunately, upon trying to figure out why it seems it has been always that way [since jbuilder days](https://github.com/ocaml/dune/commit/cdcd7e907f949fae66a0e57e569c7e0d46f40c3b).

Reviewer 3: I don't see the point in the warning. If the code stop compiling due to a missing type annotation, we just add it. A warning to pre-empt that is just noise.

Reviewer 2: Ah ok I had it backwards nevermind.

Good:

.. code:: ocaml

>     let result : A.t = A.b () in
>     let field : B.t = result.field in
>     match field with
>     | Constructor -> ...

Reviewer 2: ```suggestion
    let result : A.t = A.b () in
    match (result.field : B.t) with
    | Constructor -> ...
```
Seems more common to me.

> - When constructing records, use the qualified names in in the record, do not

Reviewer: ```suggestion
- When constructing records, use the qualified names in in the record. Do not
```

  open the record. The local open syntax pulls in all kinds of names from the
  opened module and might shadow the values that you're trying to put into the

>   record, leading to hard to debug issues.

Reviewer: ```suggestion
  record, leading to difficult debugging.
```
